src/predict.jl

```diff
@@ -0,0 +1,292 @@
+"""
+    StatsBase.predict!(y::AbstractVector{T}, m::LinearMixedModel{T},
+                    newX::AbstractMatrix{T}=m.X; use_re=true)
+    StatsBase.predict(m::LinearMixedModel{T},
+                    newX::AbstractMatrix{T}=m.X; use_re=true)
+    StatsBase.predict!(y::AbstractVector{T}, m::GeneralizedLinearMixedModel{T},
+                    newX::AbstractMatrix{T}=m.X; use_re=true, type=:response)
+    StatsBase.predict(m::GeneralizedLinearMixedModel{T},
+                    newX::AbstractMatrix{T}=m.X; use_re=true, type=:response)
+
+Predict response for new values of the fixed-effects matrix X.
+
+The modifying methods overwrite `y` in place, while the non modifying methods
+allocate a new `y`. Predictions based purely on the fixed effects can be
+obtained with `use_re=false`. In the future, it may be possible to specify
+a subset of the grouping variables to use, but not at this time.
+
+For `GeneralizedLinearMixedModel`, the `type` parameter specifies
+whether the predictions should be returned on the scale of linear predictor
+(`:linpred`) or on the response scale (`:response`). If you don't know the
+difference between these terms, then you probably want `type=:response`.
+
+!!! warning
+    Models are assumed to be full rank.
+
+!!! note
+    The `predict` and `predict!` methods with `newX` as a fixed effects matrix
+    differ from the methods with `newdata` in tabular format. The matrix
+    methods are more efficient than the tabular methods. The tabular methods
+    can accomodate new values of the grouping variable(s), but the matrix methods
+    cannot.
+"""
+function StatsBase.predict!(y::AbstractVector{<:Union{T, Missing}}, m::LinearMixedModel{T}, newX::AbstractMatrix{T}=m.X;
+                           use_re=true) where T
+    # this is called `use_re` in case we later decide to support prediction
+    # with only a subset of the RE
+
+    mul!(y, newX, m.β)
+
+    use_re || return y
+
+    # because we're using ranef(), this actually
+    # adds the *scaled* random effects
+    for (rt, bb) in zip(m.reterms, ranef(m))
+        unscaledre!(y, rt, bb)
+    end
```

Comment:
Do we know that the `reterms` will line up with the `newX`?  It seems that if `size(newX, 1) == length(y)` and `size(newX, 2) == length(m.β)` then the `use_re = false` version works but that doesn't mean that the random-effects terms will match with `length(y)`.

Reply:
It took me a minute to understand the pathology you're thinking of .... and I don't know. Looking at these now, I'm wondering whether the methods with `newX::Matrix` should just be internal. They seem like a nasty foot gun for novice users (because there's no checking that the new fixed effects matrix has columns that correspond to the coefficients in the model and you have to construct that matrix somehow) and I think it's less useful overall, except for their internal use. (Looking at the code, they are clearly internal methods that I exposed for reasons lost to me now.) In their internal use, the error checking for dimensionality matching happens before the methods are called / happens via construction of the new model, so then the problem goes away. 

The only use case this doesn't cover is predictions based purely on the fixed effects, but this is trivial for someone with tabular new data to do -- they just use the tabular methods, set `new_re_levels=:population` and use new levels of the grouping variable.

Reply:
Actually having said that, I think we can drop `predict(..., newX::Matrix)` completely -- looks like they're not even used internally in a meaningful way and can just be replaced with `mul!(y, X, coef(m))` in the appropriate place.

Reply:
Seems like a good approach.  I am inclined to avoid passing around model matrices in general as there are just too many ways to botch the process of generating them.